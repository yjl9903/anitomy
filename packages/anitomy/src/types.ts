import type { ElementCategory } from './element';

export interface AnitomyOptions {
  delimiters: string;

  parseEpisodeNumber: boolean;

  parseEpisodeTitle: boolean;

  parseFileExtension: boolean;

  parseReleaseGroup: boolean;
}

export type ParsedResult = Partial<Record<ElementCategory, string>>;

export interface AnitomyResult {
  title: string | undefined;

  type: string | undefined;

  season: string | undefined;

  year: number | undefined;

  month: number | undefined;

  language: string | undefined;

  subtitles: string | undefined;

  source: string | undefined;

  episode: {
    number: number | undefined;

    numberAlt: number | undefined;

    title: string | undefined;
  };

  volume: {
    number: number | undefined;
  };

  video: {
    term: string | undefined;

    resolution: string | undefined;
  };

  audio: {
    term: string | undefined;
  };

  release: {
    version: number | undefined;
    group: string | undefined;
  };

  file: {
    name: string;

    checksum: string | undefined;

    extension: string | undefined;
  };

  // prefix: {
  //   season: string | undefined;
  //   volume: string | undefined;
  //   episode: string | undefined;
  // };
}
