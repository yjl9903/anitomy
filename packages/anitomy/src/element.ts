export enum ElementCategory {
  AnimeSeason = 'season',
  AnimeSeasonPrefix = 'prefix.season',
  AnimeTitle = 'title',
  AnimeType = 'type',
  AnimeYear = 'year',
  AnimeMonth = 'month', // Added by https://github.com/yjl9903/AnimeGarden
  DeviceCompatibility = 'DeviceCompatibility',
  Source = 'source',
  // Episode
  EpisodeNumber = 'episode.number',
  EpisodeNumberAlt = 'episode.numberAlt',
  EpisodePrefix = 'prefix.episode',
  EpisodeTitle = 'episode.title',
  // File
  FileChecksum = 'checksum',
  FileExtension = 'extension',
  FileName = 'filename',
  // Language
  Language = 'language',
  Subtitles = 'subtitles',
  // Audio and Video
  AudioTerm = 'audio.term',
  VideoResolution = 'video.resolution',
  VideoTerm = 'video.term',
  // Volume
  VolumeNumber = 'volume',
  VolumePrefix = 'prefix.volume',
  // Release
  ReleaseGroup = 'release.group',
  ReleaseInformation = 'release.information',
  ReleaseVersion = 'release.version',
  // Other
  Unknown = 'unknown',
  Other = 'other'
}
