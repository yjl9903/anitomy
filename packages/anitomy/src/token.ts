export class TextRange {
  public text: string;

  public offset: number;

  public size: number;

  public constructor(text: string, offset: number, size: number) {
    this.text = text;
    this.offset = offset;
    this.size = size;
  }

  public fork(offset: number, size: number) {
    return new TextRange(this.text, offset, size);
  }

  public toString() {
    return this.text.slice(this.offset, this.offset + this.size);
  }
}

export enum TokenCategory {
  Unknown = 'Unknown',
  Bracket = 'Bracket',
  Delimiter = 'Delimiter',
  Identifier = 'Identifier',
  Invalid = 'Invalid'
}

export enum TokenFlag {
  // None
  None,
  // Categories
  Bracket,
  NotBracket,
  Delimiter,
  NotDelimiter,
  Identifier,
  NotIdentifier,
  Unknown,
  NotUnknown,
  Valid,
  NotValid,
  // Enclosed (Meaning that it is enclosed in some bracket (e.g. [ ] ))
  Enclosed,
  NotEnclosed
}

export interface Token {
  category: TokenCategory;

  content: string;

  enclosed: boolean;
}

function checkTokenFlags(token: Token, flags: TokenFlag[]) {
  // Make sure token is the correct closure
  if (flags.some((f) => f === TokenFlag.Enclosed || f === TokenFlag.NotEnclosed)) {
    const success = flags.includes(TokenFlag.Enclosed) === token.enclosed;
    if (!success) return false; // Not enclosed correctly (e.g. enclosed when we're looking for non-enclosed).
  }

  // Make sure token is the correct category
  if (!flags.some((f) => TokenFlag.Bracket <= f && f <= TokenFlag.NotValid)) {
    return true;
  }

  const tasks: [TokenFlag, TokenFlag, TokenCategory][] = [
    [TokenFlag.Bracket, TokenFlag.NotBracket, TokenCategory.Bracket],
    [TokenFlag.Delimiter, TokenFlag.NotDelimiter, TokenCategory.Delimiter],
    [TokenFlag.Identifier, TokenFlag.NotIdentifier, TokenCategory.Identifier],
    [TokenFlag.Unknown, TokenFlag.NotUnknown, TokenCategory.Unknown],
    [TokenFlag.NotValid, TokenFlag.Valid, TokenCategory.Invalid]
  ];
  for (const [fe, fn, c] of tasks) {
    const success = flags.includes(fe)
      ? token.category === c
      : flags.includes(fn) && token.category !== c;
    if (success) return true;
  }
  return false;
}

export function findNextToken(tokens: Token[], position: number, ...flags: TokenFlag[]) {
  for (let i = position + 1; i < tokens.length; i++) {
    if (tokens[i] && checkTokenFlags(tokens[i], flags)) {
      return i;
    }
  }
  return tokens.length;
}

export function findPrevToken(tokens: Token[], position: number, ...flags: TokenFlag[]) {
  for (let i = position - 1; i >= 0; i--) {
    if (tokens[i] && checkTokenFlags(tokens[i], flags)) {
      return i;
    }
  }
  return -1;
}

export function findToken(tokens: Token[], start: number, end: number, ...flags: TokenFlag[]) {
  for (let i = start; i < end; i++) {
    if (tokens[i] && checkTokenFlags(tokens[i], flags)) {
      return i;
    }
  }
  return -1;
}
