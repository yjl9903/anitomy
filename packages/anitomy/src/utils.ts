import type { ParsedResult } from './types';

export function inRange<T>(list: T[], idx: number) {
  return 0 <= idx && idx < list.length;
}

export function isNumericString(text: string) {
  return /^\d+$/.test(text);
}

export function trim(text: string, removal: string[]) {
  let start = 0,
    end = text.length - 1;
  while (start <= end && removal.includes(text[start])) {
    start++;
  }
  while (end >= start && removal.includes(text[end])) {
    end--;
  }
  return text.slice(start, end + 1);
}

export function mergeResult(source: ParsedResult, income: ParsedResult = {}) {
  return {
    ...source,
    ...income
  };
}
