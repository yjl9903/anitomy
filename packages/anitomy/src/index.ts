import type { AnitomyOptions, AnitomyResult, ParsedResult } from './types';

import { mergeResult } from './utils';
import { KeywordManager } from './keyword';
import { ElementCategory } from './element';
import { parse as doParse } from './parser';
import { tokenize as doTokenize } from './tokenizer';

export type * from './types';

/**
 * Parser instance with shared options and parsed result cache
 */
export class Parser {
  private readonly options: AnitomyOptions;

  private readonly cache = new Map<string, AnitomyResult | null>();

  public constructor(options: Partial<AnitomyOptions> = {}) {
    this.options = resolveOptions(options);
  }

  public parse(filename: string, force = false) {
    if (!force && this.cache.has(filename)) {
      const result = this.cache.get(filename);
      return result ? result : undefined;
    } else {
      const result = parse(filename, this.options);
      if (result) {
        this.cache.set(filename, result);
      } else {
        this.cache.set(filename, null);
      }
      return result;
    }
  }
}

export function parse(
  filename: string,
  _options: Partial<AnitomyOptions> = {}
): AnitomyResult | undefined {
  if (filename === '') return undefined;

  let result: ParsedResult = {};

  const options = resolveOptions(_options);

  result.filename = filename;
  if (options.parseFileExtension) {
    const ext = removeExtension(filename);
    if (ext) {
      result.filename = filename;
      result.extension = ext.extension;
    }
  }

  const tokenized = doTokenize(result.filename!, options);
  result = mergeResult(result, tokenized.result);
  if (!tokenized.ok) {
    return resolveResult(result);
  }

  const parsed = doParse(result, tokenized.tokens, options);
  result = parsed.result;

  return resolveResult(result);
}

export function resolveOptions(options: Partial<AnitomyOptions>): AnitomyOptions {
  return {
    delimiters: ' _.&+,|',
    parseEpisodeNumber: true,
    parseEpisodeTitle: true,
    parseFileExtension: true,
    parseReleaseGroup: true,
    ...options
  };
}

function resolveResult(result: ParsedResult): AnitomyResult {
  const resolved: AnitomyResult = {
    title: result['title'],
    type: result['type'],
    season: normalizeSeason(result['season']),
    year: normalizeNumber(result['year']),
    month: normalizeNumber(result['month']),
    language: result['language'],
    subtitles: result['subtitles'],
    source: result['source'],
    episode: {
      number: normalizeNumber(result['episode.number']),
      numberAlt: normalizeNumber(result['episode.numberAlt']),
      title: result['episode.title']
    },
    volume: {
      number: normalizeNumber(result['volume'])
    },
    video: {
      term: result['video.term'],
      resolution: result['video.resolution']
    },
    audio: {
      term: result['audio.term']
    },
    release: {
      version: normalizeNumber(result['release.version']),
      group: result['release.group']
    },
    file: {
      name: result['filename']!,
      extension: result['extension'],
      checksum: result['checksum']
    }
  };
  return resolved;

  function normalizeSeason(num: string | undefined) {
    if (num !== undefined && num !== null) {
      return /^\d+$/.test(num) ? String(+num) : num;
    } else {
      return undefined;
    }
  }

  function normalizeNumber(num: string | undefined) {
    try {
      if (num !== undefined && num !== null) {
        const n = parseFloat(num);
        return !Number.isNaN(n) ? n : undefined;
      }
    } catch {}
    return undefined;
  }
}

function removeExtension(filename: string) {
  const position = filename.lastIndexOf('.');
  if (position === -1) return undefined;

  const extension = filename.slice(position + 1);
  if (extension.length > 4 || !/^[a-zA-Z0-0]+$/.test(extension)) {
    return undefined;
  }

  if (!KeywordManager.contains(ElementCategory.FileExtension, extension)) {
    return undefined;
  }

  return {
    filename: filename.slice(0, position),
    extension
  };
}
