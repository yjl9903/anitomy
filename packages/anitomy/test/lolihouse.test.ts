import { describe, it, expect } from 'vitest';

import { parse } from '../src';

import { LoliHouse } from './testcase';

describe('LoliHouse', () => {
  it('should parse', () => {
    for (const filename of LoliHouse) {
      const info = parse(filename);
      expect(info).toMatchSnapshot();
      expect(info?.title).toBeTruthy();
    }
  });
});
