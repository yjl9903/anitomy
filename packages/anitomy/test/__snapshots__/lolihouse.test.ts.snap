// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoliHouse > should parse 1`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[LoliHouse] 肌肉魔法使-MASHLE- / Mashle - 02 [WebRip 1080p HEVC-10bit AAC][简繁内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "肌肉魔法使-MASHLE- / Mashle",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 2`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": undefined,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[LoliHouse] 赤色光弹Zillion / Akai Koudan Zillion [BDRip 720p HEVC-10bit AAC]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "BDRip",
  "subtitles": undefined,
  "title": "赤色光弹Zillion / Akai Koudan Zillion",
  "type": undefined,
  "video": {
    "resolution": "720p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 3`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 10,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[喵萌Production&LoliHouse] UniteUp! 众星齐聚 - 10 [WebRip 1080p HEVC-10bit AAC][简繁日内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "UniteUp! 众星齐聚",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 4`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[豌豆字幕组&风之圣殿字幕组&LoliHouse] 新石纪 第三季 / 石纪元 第三季 / Dr.STONE S3 - 02 [WebRip 1080p HEVC-10bit AAC][简繁外挂字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": "3",
  "source": "WebRip",
  "subtitles": undefined,
  "title": "新石纪 第三季 / 石纪元 第三季 / Dr STONE S3",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 5`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[喵萌Production&LoliHouse] 偶像大师 灰姑娘女孩 U149 / THE IDOLM@STER CINDERELLA GIRLS U149 - 02 [WebRip 1080p HEVC-10bit AAC][简繁日内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "偶像大师 灰姑娘女孩 U149 / THE IDOLM@STER CINDERELLA GIRLS U149",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 6`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[喵萌奶茶屋&LoliHouse] 跃动青春 / Skip to Loafer - 02 [WebRip 1080p HEVC-10bit AAC][简繁日内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "跃动青春 / Skip to Loafer",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 7`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": 12,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[动漫国字幕组&LoliHouse] 关于我在无意间被隔壁的天使变成废柴这件事 / Otonari no Tenshi-sama [01-12 修正合集] [WebRip 1080p HEVC-10bit AAC][简繁内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "关于我在无意间被隔壁的天使变成废柴这件事 / Otonari no Tenshi-sama",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 8`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[喵萌奶茶屋&LoliHouse] 放学后失眠的你 / Kimi wa Houkago Insomnia - 01 [WebRip 1080p HEVC-10bit AAC][简繁日内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "放学后失眠的你 / Kimi wa Houkago Insomnia",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 9`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[喵萌奶茶屋&LoliHouse] 勇者死了！/ Yuusha ga Shinda! - 01 [WebRip 1080p HEVC-10bit AAC][简繁内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "勇者死了！/ Yuusha ga Shinda!",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 10`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[动漫国字幕组&LoliHouse] 魔法使的新娘 第二季 / Mahou Tsukai no Yome S2 - 01 [WebRip 1080p HEVC-10bit AAC][简繁内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": "2",
  "source": "WebRip",
  "subtitles": undefined,
  "title": "魔法使的新娘 第二季 / Mahou Tsukai no Yome S2",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 11`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[动漫国字幕组&LoliHouse] 为美好的世界献上爆焰! / Kono Subarashii Sekai ni Bakuen wo! - 01 [WebRip 1080p HEVC-10bit AAC][简繁内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "为美好的世界献上爆焰! / Kono Subarashii Sekai ni Bakuen wo!",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 12`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[喵萌奶茶屋&LoliHouse] 与山田谈一场Lv999的恋爱 / Yamada-kun to Lv999 no Koi wo Suru - 02 [WebRip 1080p HEVC-10bit AAC][简繁日内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "与山田谈一场Lv999的恋爱 / Yamada-kun to Lv999 no Koi wo Suru",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 13`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 12,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[悠哈璃羽字幕社&LoliHouse] 冰属性男子与无表情女子 / Koori Zokusei Danshi to Cool na Douryou Joshi - 12 [WebRip 1080p HEVC-10bit AAC][简繁内封字幕][END]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "冰属性男子与无表情女子 / Koori Zokusei Danshi to Cool na Douryou Joshi",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 14`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 12,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[悠哈璃羽字幕社&LoliHouse] 不相信人类的冒险者们好像要去拯救世界 / Ningen Fushin - 12 [WebRip 1080p HEVC-10bit AAC][简繁内封字幕][END]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "不相信人类的冒险者们好像要去拯救世界 / Ningen Fushin",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 15`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[LoliHouse] 亡骸游戏 / 尸体如山的死亡游戏 / Dead Mount Death Play - 01 [WebRip 1080p HEVC-10bit AAC][简繁内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "亡骸游戏 / 尸体如山的死亡游戏 / Dead Mount Death Play",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 16`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[喵萌奶茶屋&LoliHouse] 女神咖啡厅 / 女神的露天咖啡厅 / Megami no Cafe Terrace - 01 [WebRip 1080p HEVC-10bit AAC][简繁日内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "女神咖啡厅 / 女神的露天咖啡厅 / Megami no Cafe Terrace",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 17`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[喵萌Production&LoliHouse] 世界大明星 / World Dai Star - 01 [WebRip 1080p HEVC-10bit AAC][简繁日内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "世界大明星 / World Dai Star",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 18`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[喵萌奶茶屋&LoliHouse] 转生贵族的异世界冒险录 ~不知克制的众神的使徒~ / Tensei Kizoku no Isekai Boukenroku - 02 [WebRip 1080p HEVC-10bit AAC][简繁日内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "转生贵族的异世界冒险录 ~不知克制的众神的使徒~ / Tensei Kizoku no Isekai Boukenroku",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 19`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 45,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[豌豆字幕组&风之圣殿字幕组&LoliHouse] 鬼灭之刃 刀匠村篇 01(45) / Kimetsu no Yaiba - 45 [WebRip 1080p HEVC-10bit AAC][简繁外挂字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "鬼灭之刃 刀匠村篇 01",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 20`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 27,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[动漫国字幕组&LoliHouse] 伊甸星原 / EDENS ZERO - 27 [WebRip 1080p HEVC-10bit AAC][简繁内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "伊甸星原 / EDENS ZERO",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 21`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[喵萌奶茶屋&LoliHouse] 我内心的糟糕念头 / Boku no Kokoro no Yabai Yatsu - 02 [WebRip 1080p HEVC-10bit AAC][简繁日内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "我内心的糟糕念头 / Boku no Kokoro no Yabai Yatsu",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 22`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 13,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[喵萌奶茶屋&LoliHouse] 机动战士高达 水星的魔女 / Mobile Suit Gundam THE WITCH FROM MERCURY - 13 [WebRip 1080p HEVC-10bit AAC][简繁日内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "机动战士高达 水星的魔女 / Mobile Suit Gundam THE WITCH FROM MERCURY",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 23`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[LoliHouse] 第二次被异世界召唤 / 異世界召喚は二度目です / Isekai Shoukan wa Nidome desu - 01 [WebRip 1080p HEVC-10bit AAC][简繁内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "第二次被异世界召唤 / 異世界召喚は二度目です / Isekai Shoukan wa Nidome desu",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`LoliHouse > should parse 24`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[SweetSub&LoliHouse] 天国大魔境 / Tengoku Daimakyou - 02 [WebRip 1080p HEVC-10bit AAC][简繁日内封字幕]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "天国大魔境 / Tengoku Daimakyou",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;
